samza-kv/src/test/scala/org/apache/samza/storage/kv/TestLargeMessageSafeStore.scala

```diff
@@ -0,0 +1,187 @@
+
+  val store = mock(classOf[KeyValueStore[Array[Byte], Array[Byte]]])
+  val storeName = "testStore"
+  val storeConfig = mock(classOf[Config])
+  val systemStreamPartition = mock(classOf[SystemStreamPartition])
+  val collector = mock(classOf[MessageCollector])
+  val metrics = mock(classOf[LoggedStoreMetrics])
+  val maxMessageSize = 1024
+  val counter = mock(classOf[Counter])
+
+  @Before
+  def setup: Unit = {
+    when(systemStreamPartition.getPartition).thenReturn(new Partition(1))
+    when(storeConfig.getInt(Matchers.eq("changelog.max.message.size.bytes"), Matchers.any())).thenReturn(maxMessageSize)
+    when(metrics.puts).thenReturn(counter)
+    when(counter.inc()).thenReturn(1)
+    doNothing().when(collector).send(Matchers.any())
+    doNothing().when(store).put(Matchers.any(), Matchers.any())
+    doNothing().when(store).putAll(Matchers.any())
+  }
+
+  @Test
+  def testLargeMessagePutWithDropLargeMessageDisabled = {
+    val largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize)
+
+    val key1 = new Array[Byte](16)
+    val smallMessage = new Array[Byte](32)
+    largeMessageSafeKeyValueStore.put(key1, smallMessage)
+    verify(store, times(1)).put(Matchers.eq(key1), Matchers.eq(smallMessage))
+
+    val key2 = new Array[Byte](16)
+    val largeMessage = new Array[Byte](maxMessageSize + 1)
+    try {
+      largeMessageSafeKeyValueStore.put(key2, largeMessage)
+      fail("The test case should have failed due to a large message being passed to the changelog, but it didn't.")
+    } catch {
+      case e: SamzaException =>
+        verify(store, times(0)).put(Matchers.eq(key2), Matchers.eq(largeMessage))
+    }
+  }
+
+  @Test
+  def testSmallMessagePutAllSuccessWithDropLargeMessageDisabled = {
+    val largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize)
+    val key = new Array[Byte](16)
+    val smallMessage = new Array[Byte](32)
+
+    val entries = new ArrayList[Entry[Array[Byte], Array[Byte]]]
+    entries.add(new Entry(key, smallMessage))
+
+    largeMessageSafeKeyValueStore.putAll(entries)
+    verify(store, times(1)).putAll(Matchers.eq(entries))
+    verify(store, times(0)).put(Matchers.any(classOf[Array[Byte]]), Matchers.any(classOf[Array[Byte]]))
+  }
+
+  @Test
+  def testLargeMessagePutAllFailureWithDropLargeMessageDisabled = {
+    val largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize)
+    val key = new Array[Byte](16)
+    val largeMessage = new Array[Byte](maxMessageSize + 1)
+
+    val entries = new ArrayList[Entry[Array[Byte], Array[Byte]]]
+    entries.add(new Entry(key, largeMessage))
+
+    try {
+      largeMessageSafeKeyValueStore.putAll(entries)
+      fail("The test case should have failed due to a large message being passed to the changelog, but it didn't.")
+    } catch {
+      case e: SamzaException =>
+        verify(store, times(0)).putAll(Matchers.eq(entries))
```

Comment:
You can do `verifyZeroInteractions(store)` to make sure it just never got used at all.

Reply:
Done.